README.md

```diff
@@ -193,3 +193,61 @@ $ telnet 192.168.99.100 1234
 
 This `-p` argument links port 4321 on the container to port 1234 on the VM network adapter. It must appear after `docker run` and before `dmscid/plankton`. This allows us to connect to the container from outside of the VM, in this case using Telnet. The `192.168.99.100` IP is the IP of the VM on the bridge network between the host and the VM. VirtualBox will typically use this IP when available, but it may be different on your system.
 
+## Remote access to devices
+
+*Please note that this functionality should only be used on a trusted network.*
+
+Besides the device specific protocols, the device can be made accessible from the outside via JSON-RPC over ZMQ. This can be achieved by passing the `-r` option with a `host:port` string to the simulation:
+
+```
+$ python simulation.py -r 127.0.0.1:10000 -d chopper -- -p SIM:
+```
+
+Now the device can be controlled via the `control.py`-script in a different terminal window. The service can be queried to show the available objects using the `-l` or `--list-objects` flag:
+
+```
+$ python control.py --ip 127.0.0.1 --port 10000 --list-objects
+```
+
+The `ip` and `port` parameters default to the values shown here, so they will be omitted in the following examples. To get information on the API of an object, the `-a` or `--show-api` option can be used in conjunction with a device name obtained from the previous command:
+
+```
+$ python control.py -a device
+```
+
+This will output a list of properties and methods which is available for remote access. This may not comprise the full interface of the object depending on the server side configuration. Obtaining the value of a property is done like this:
+
+```
+$ python control.py device state
+```
+
+The same syntax is used to call methods without parameters:
+
+```
+$ python control.py device initialize
+```
+
+To set a property to a new value, the value has to be supplied on the command line:
+
+```
+$ python control.py device targetSpeed 100
+$ python control.py device start
+```
+
+Only numeric types and strings can be used as arguments via the `control.py`-script. The script always tries to convert parameters to `int` first, then to `float` and leaves it as `str` if both fail. For other types and more control over types, it's advised to write a Python script instead using the tools provided in `core.rpc_client` which makes it possible to use the remote objects more or less transparently. An example to control the chopper:
+
+```
+from time import sleep
+from core.rpc_client import ZMQJSONRPCConnection, get_remote_object_collection
```

Comment:
Hm... I might be misunderstanding something about how the single-object scenario works. But just looking at the way it is now:

```python
class ControlClient(object):
    def get_object(self, object_name='')
        pass

def remote_objects(source, object_name=''):
    """
    :param source: A source of remote objects, must be of type ControlClient.
    """
    if not isinstance(source, ControlClient):
        raise RuntimeError('A ControlClient instance is required to retrieve remote objects from the server side.')
```

The first argument to `remote_objects` is already necessarily a ControlClient. So why can't that be turned into a member method, with the first argument being `self`? Don't need the `isinstance` check that way, and it feels conceptually correct from an object oriented perspective (you want the remote objects from that client... `ControlClient.remote_objects()`).

For the `dict` notation, I would see that as just a convenience alias for `get_object`. So I don't see the harm in adding it (and keeping the explicit `get_object`). That way you can use whichever is more readable for what you want to do:

```python
class ControlClient(object):
    def get_object(self, object_name='')
        pass

    def remote_objects(self, object_name=''):
        object_names = self.get_object(object_name).get_objects()
        return {obj: self.get_object(obj) for obj in object_names}

    def __getitem__(self, object_name):
        return self.get_object(object_name)
```

>The other alternative I see would be that the dict like interface is just empty if the remote object is not of the right type, but at the moment that does not convince me, because you still need a way to get the top level object, so there are two interfaces in one...

You're right, of course... it's a bit awkward that sometimes the dictionary notation just doesn't make sense, depending on what the server is hosting. But isn't *that* the source of the issue? That the server is exposing things in an inconsistent way?

Can't we just say the server always hosts a collection at the top level. And if you just want to expose one object, then you just expose one object... but it still has to have a name and be inside the top level collection. What would be the downside of that?